### DISCUSSION TITLE
Dialogservice Unsubscribe

### DISCUSSION THREAD

**Discussion by arteMatth** [primefaces/community#2776] [PrimeNG]
Hello, I have a question about the Dialogservice. 

When I use the Dialogservice and subscribe to the onClose Event, do I have to destroy the subscribtion after or does it happen automatically.

Thanks in advance.

**Comment by qwadrox** [comment]
Hi @arteMatth! This is a good question.

**Short answer: yes, you should unsubscribe!**

**Long answer:** 

This is what happens internally: https://github.com/primefaces/primeng/blob/1994a1310843f5047181d19d263b5a9fb3876b62/src/app/components/dynamicdialog/dynamicdialog-ref.ts#L13-L19

After one second, the _onClose method is being completed after the emission and as we know very well:

> When an Observable issues an OnError or OnComplete notification to its observers, this ends the subscription.

So technically, the manual termination of the subscription is not required. However, there is no harm in doing so.

The main rule is, and what I recommend following in your applications:

```
In general: Don't unsubscribe when using HttpClient. It is only necessary when you have logic affecting global application state in the subscription because it eventually completes.
Everywhere else: Do unsubscribe.
```

In our case, what if the underlying logic had not included the completion of the subject? We cannot trust the library authors, as they're not robots, and they can make mistakes, as can Angular itself. We don't know when internal logic might change or malfunction. So you won't harm the application by manually unsubscribing, just in case. In this way, you don't have to think about every single time do I have to unsubscribe or not. 

There is now a very convenient way to unsubscribe with the `takeUntilDestroyed` operator: https://medium.com/@chandrashekharsingh25/exploring-the-takeuntildestroyed-operator-in-angular-d7244c24a43e


References:
https://stackoverflow.com/questions/41334931/do-we-need-to-unsubscribe-from-observable-that-completes-errors-out
https://github.com/ueler/angular-rxjs-unsubscribe

**Reply by arteMatth** [reply]
Thank you very much for this detailed explanation. 

It makes a lot of sense and is easy to understand. 

As you mentioned, I will additionally unsubscribe the dialog, just to be on the save side.
